### ISSUE TITLE
Project Image to W

### ISSUE THREAD

**Issue by Miriam2040** [NVlabs/stylegan3#136]
If I am creating image by style gan can I project it to same W that originally it was generated by? If so how?

**Comment by PDillis** [comment]
It depends on what you want to do, as I don't quite understand what you want: do you want the dlatent/w vector that generates a random image or is the image originally generated by the Generator?

In either case, you can use the projection code from [StyleGAN2](https://github.com/NVlabs/stylegan2-ada-pytorch/blob/main/projector.py). Note that if your model is a StyleGAN3 one (`config-t` or `config-r`), then the image might be rotated/translated and the final result won't be very good. You can mitigate this by anchoring the latent space to `w_avg`. An example of this can be found [here](https://github.com/PDillis/stylegan3-fun/blob/4ce9d6f7601641ba1e2906ed97f2739a63fb96e2/torch_utils/gen_utils.py#L334).

Now, if you originally generated the image with the Generator and just want to save the respective dlatent, then this is straightforward. An example on how to do this using a seed is like so:

```py
seed = 0  # Random seed to be used
truncation_psi = 1.0  # Whatever truncation psi you used
noise_mode = 'const'  # Noise mode for the synthesis block
run_dir = 'out'  # Wherever you want to save the image; best to automate this as in the training loop

# Get the dlatent
z = np.random.RandomState(seed).randn(1, G.z_dim)
w = G.mapping(torch.from_numpy(z).to(device), None)
w_avg = G.mapping.w_avg
w = w_avg + (w - w_avg) * truncation_psi

# Generate the image
synth_image  = G.synthesis(w, noise_mode=noise_mode)
synth_image = (synth_image + 1) * 255 / 2  # [-1.0, 1.0] => [0.0, 255.0]
synth_image = synth_image.permute(0, 2, 3, 1).clamp(0, 255).to(torch.uint8).cpu().numpy()  # NCHW => NWHC and [0.0, 255.0] => [0, 255]

# Save both the dlatent and image
np.save(os.path.join(run_dir, 'image.npy'), w.unsqueeze(0).cpu().numpy())
PIL.Image.fromarray(synth_image, 'RGB').save(os.path.join(run_dir, 'image.jpg'))
```

Best of luck!
